### ISSUE TITLE
OpenGlBufferQtQuick OpenGL items can't be rendered

### ISSUE THREAD

**Issue by lucaszanella** [status-im/react-native-desktop#462]
I'm trying to return my custom QQuickItem object, which is a subclass of OpenGlBufferQtQuick. However, I must return it as a QQuickItem. Therefore, when it reaches the screen, its crucial items aren't called, so nothing is rendered.

I'm returning a https://github.com/lucaszanella/orwell/blob/57a2dab0dd770dde7f8c96ed3f7ea8672cc88f81/orwellJS/modules/mediaplayer/desktop/mediaplayermanager.cpp#L42 which can be seen declared here: https://github.com/lucaszanella/orwell/blob/57a2dab0dd770dde7f8c96ed3f7ea8672cc88f81/OpenGlBufferQtQuick.cpp

It looks like it is not possible to render OpenGL the right way, I guess. I think something can be added to call OpenGlBufferQtQuick's methods when its placed on the screen.

Can you understand the problem I'm having?

**Comment by vkjr** [comment]
@lucaszanella, If you suspect that there is an issue in `react-native-desktop`, please prepare minimal example project that showcases the problem.
You can take [this minimal project](https://github.com/vkjr/BezierCurveRnd) as a start point and modify it by changing BezierCurve to your control and check if it still fails.

**Comment by vkjr** [comment]
Also make sure you set some size to your created widget like it is done in example project. It could be that you don’t see anything because wodget is 0x0 size.

**Comment by lucaszanella** [comment]
@vkjr you're right. I now see the widget in react-native-desktop if I set the width. However I didn't need to set the width for my old QQuickItem, just this one. That's odd. Anyways, I used

```
QQuickItem* MediaPlayerManager::createView(const QVariantMap& properties) {
    OpenGlBufferItem* item = new OpenGlBufferItem();
    item->setWidth(50);
    item->setHeight(22);
    std::cout << "returning custom item" << std::endl;
    return item;
}
```
However its size in the react-native-desktop is the entire screen, while if the item is called using QML (no react native at all) it'll have the right dimensions specified in QML. I think it's because I didn't make it my `OpenGlBufferQtQuick` a subclass of `ReactItem` yet. I don't know if it's possible because both `ReactItem` and `OpenGlBufferQtQuick` inherit from `QQuickItem`, so we'll end up with the diamond problem. And I can't alter `OpenGlBufferQtQuick` neither `ReactItem` because one is from QT and the other is from React Native Desktop.

**Comment by vkjr** [comment]
ReacItem provides flexbox capabilities. You don't have to subclass it to have `setWidth()` and `setHeight()` calls working.

If you want to manage item size from javascript I can suggest following:

**In js nest your component under <View>:**
```
<View style={{width: 200, height: 100}}>
  < OpenGlBufferQtQuick/>
</View>
```

**In OpenGlBufferQtQuick.cpp**
you can make your newly created item to follow parent size. Override `onItemChange` to detect when parent is set and add connections like in [this link](https://stackoverflow.com/questions/34721748/is-it-possible-to-access-qml-anchors-from-c) (last answer)

This way you'll be able to set size to <View> and it will be picked by OpenGlBufferQtQuick

**Comment by lucaszanella** [comment]
if I add

`    explicit OpenGlBufferItem(QQuickItem* parent){`

then I cannot use 

```
QQuickItem* MediaPlayerManager::createView(const QVariantMap& properties) {
    OpenGlBufferItem* item = new OpenGlBufferItem();
```
to return it, because I don't have a QQuickItem* parent.

Without using this, I tried what you said about putting things inside a View, that's the result:

https://imgur.com/a/4DCDbTs

As you can see, it did set the image's size, but in front of another image with the full size. Maybe I get two images because I didn't set the widthChanged and heightChanded signals?

Abut the setWidth and setHeight: you said it should work without subclassing ReactItem. Since mine does not work, does that mean I did something wrong? 

I don't want to do things the wrong way. I think that putting every camera stream inside a view is not right :(

also: Is it possible to set properties like width height and custom ones like this:

<OrwellMediaPlayer width=200 height=100 uri="rtsp://mycamera:10554"/> 

?

**Comment by lucaszanella** [comment]
I was talking about this: https://stackoverflow.com/questions/55210837/solving-this-specific-c-diamond-problem-for-qt-classes

looks like  OpenGlBufferQtQuick OpenGL items really can't be rendered

**Comment by vkjr** [comment]
@lucaszanella, as I suggested in another discussion thread, please create a minimal example that reproduces your issue with full description of what you are trying to achieve and post it. Then I'll be able to run it and take a look. That would be much quicker.

**Comment by lattice0** [comment]
Here's the minimal example.

I return an `OpenGlBufferItem` because it is a `QQuickItem`:

However, as you can see here:

it's a `QQuickFramebufferObject` which inherits from `QQuickItem`. In order for it to be a React Native Item, I have to make it inherit `ReactItem`, which inherits from `QQuickItem` also. Therefore it's impossible to make my `OpenGlBufferItem` implement `ReactItem` because we'd end up with the diamond problem.

We have the following

               QQuickItem
              /          \
             /            \
    QQuickPaintedItem QQuickFramebufferObject
           /                \
       ReactItem        OpenGlBufferItem

What I need is

               QQuickItem
              /          \
             /            \
    QQuickPaintedItem QQuickFramebufferObject
           /                \
       ReactItem            /
           \               /
            \             /
            OpenGlBufferItem

which is impossible
